File: ReadyTasks/FSharp/ProjectEuler/Task_4/Task_4/Program.fs
open System

let getReverseNumb n =
    let rec helpFunc numb res =
        if (numb = 0) then
            res
        else
            helpFunc (numb / 10) (res * 10 + numb % 10)
    helpFunc n 0

let isPallindrome numb =
    numb = (getReverseNumb numb)

let getMaxPallindromeProduct digitNumb =
    let min = pown 10 digitNumb
    let max = (pown 10 (digitNumb + 1)) - 1
    [min..max] |> List.allPairs [min..max] |> List.map (fun (a,b) -> a * b) |> List.filter isPallindrome |> List.max

[<EntryPoint>]
let main argv =
    let result = getMaxPallindromeProduct 2
    printfn "%d" result
    0 // return an integer exit code

File: ReadyTasks/CSharp/ProjectEuler/Task_4/Task_4/Program.cs
using System;

namespace Task_4
{
    class Program
    {
        static int GetReverseNumber(int numb)
        {
            int result = 0;
            while(numb > 0)
            {
                result *= 10;
                result += numb % 10;
                numb /= 10;
            }
            return result;
        }

        static bool IsPallindrome(int numb)
        {
            return numb == GetReverseNumber(numb);
        }

        static int GetMaxProduct(int max1, int max2)
        {
            int result = 0;
            for (int i = 1; i <= max1; i++)
            {
                for (int j = 1; j <= max2; j++)
                {
                    int product = i * j;
                    if (IsPallindrome(product))
                    {
                        result = Math.Max(result, product);
                    }
                }
            }
            return result;
        }

        static void Main(string[] args)
        {
            Console.WriteLine(GetMaxProduct(999, 999));
        }
    }
}
